---
title: "Convergence analysis among monophyletic pairs with contrasting GC phenotype"
author: "Stefano Giulieri, Romain Guerillot"
date: "25/02/2021"
output: html_document
editor_options: 
  chunk_output_type: console
---

Here we check for convergent mutations / convergently mutated genes in carefully created monophyletic pairs within the VANANZ dataset. Each cluster consists of a pair with high growth rate (iso1) and one with low growth (iso2).

To get mutations arising in low GC isolates (SCV-like), we ran snippy within the pair, where the de novo assembly of iso1 is the reference.

```{r setup, include=F}
knitr::opts_chunk$set(echo = TRUE)
knitr::opts_knit$set(root.dir = here::here())
setwd(here::here())
print(paste("My working directory is:" ,here::here()))
```

```{r message=F}
library(tidyverse)
library(ggtree)
library(phytools)
rm(list = ls())
```

# Import raw data

Processed snippy output

```{r}
df_snippy <- readRDS("Genetic_pairs_analysis_Oct_2020/processed_data/snippy/denovo/snippy_annotated.Rda") %>%
  mutate(pair_id = paste0(iso1, "--", iso2))
```

Dataset of contrasting GC uptake

```{r}
df_pair_GC <- read_csv("Genetic_pairs_analysis_Oct_2020/processed_data/monophyletic_clusters/genetic_pairs_significant_GC_change.csv")
```

# Check pairs genetic distance and significance directionality of GC pairs (AUC GC iso1 > AUC GC iso2)
```{r}
df_pair_GC <- readRDS("Genetic_pairs_analysis_Oct_2020/processed_data/GC/dataframes/parameters/genetic_pairs_AUC_growth_stats.Rda") %>%
  merge(df_pair_GC, ., by = c("iso1", "iso2"), all.x = T)
```

Difference is not significant for 11 pairs. We remove pairs with p>=0.05

```{r}
# df_pair_GC %>%
#   filter(delta_AUC_growth_pval > 0.05) %>%
#   .$pair_id
# 
# df_pair_GC <- df_pair_GC %>%
#   filter(delta_AUC_growth_pval < 0.05)
```

distance range:
- de novo mapping 0-171
- core snp 0-158

# Filter snippy output based on clusters

Check that all pairs are present in the snippy output file

```{r}
missing_pairs <- setdiff(df_pair_GC$pair_id, df_snippy$pair_id)
missing_pairs
```

These pairs could be missing because 1) no mutations were found or 2) they were not including in the genetic pairs, due to an artifact of the distance measured using the core genome.

```{r}
df_pair_GC %>%
  filter(pair_id %in% missing_pairs) %>%
  select(contains("dist"))
```

The 4 pairs missing from snippy have 0 mutations. These could be interesting pairs for long reads sequencing or at least they should be checked for structural variants based on short reads data!

# Convergence among 21 pairs (17 with mutations detected)

```{r}
df_snippy_pair_GC <- df_pair_GC %>%
  left_join(df_snippy) %>%
  drop_na(CHROM)
 
n_distinct(df_snippy_pair_GC$pair_id) 
```

# Convergence?

First we mask intergenic mutations and synonymous mutations.

```{r}
df_snippy_pair_GC %>%
  count(EFFTYPE)
df_snippy_prot_changes <- df_snippy_pair_GC %>%
  drop_na(cdhit_group) %>%
  filter(!is.na(EFFTYPE) & EFFTYPE != "synonymous_variant")
```

Check for convergent *mutations*: same gene, same mutation

```{r}
df_converg_mutations <- df_snippy_prot_changes %>%
  group_by(cdhit_group, MUTATION_SHORT ) %>%
  mutate(n_events = n_distinct(pair_id)) %>%
  relocate(n_events)

df_converg_mutations %>%
  filter(n_events > 1) %>%
  distinct(pair_id, cdhit_group, pan_gene_symbol, neb_locus_tag, neb_mutant_id, neb_product, MUTATION_SHORT) %>%
  arrange(cdhit_group)
```

We have 0 convergent identical mutations

Check for convergent *positions*: same gene, same position

```{r}
df_converg_positions <- df_snippy_prot_changes %>%
  group_by(cdhit_group, AA_POS ) %>%
  mutate(n_events = n_distinct(pair_id)) %>%
  relocate(n_events)

df_converg_positions %>%
  filter(n_events > 1) %>%
  distinct(pair_id, cdhit_group, pan_gene_symbol, neb_locus_tag, neb_mutant_id, neb_product, MUTATION_SHORT) %>%
  arrange(cdhit_group)
```


We have 9 mutations that are shared between two pairs. This raises the question as to whether these two pairs are truly independent.

We will remove one of the pair between the pairs BPH3258--BPH3276  / BPH3708--BPH3541 to insure that all pairs are independent.

```{r}
df_pair_GC %>% filter(pair_id %in% c("BPH3258--BPH3276", "BPH3708--BPH3541"))
```

We will remove the pair BPH3708 BPH3541 because of both superior delta AUC GC and inferior genetic distance when compared with their potentially redundant pairs

Uptade mutation dataframe and
Recheck for convergent *mutations*: same gene, same mutation


```{r}
df_cleaned_pair_GC <- df_pair_GC %>%
  filter(!pair_id %in% c("BPH3708--BPH3541"))

df_snippy_prot_changes <- df_snippy_pair_GC %>%
  filter(!pair_id %in% c("BPH3708--BPH3541")) %>%
  drop_na(cdhit_group) %>%
  filter(!is.na(EFFTYPE) & EFFTYPE != "synonymous_variant")


df_converg_mutations <- df_snippy_prot_changes %>%
  group_by(cdhit_group, MUTATION_SHORT ) %>%
  mutate(n_events = n_distinct(pair_id)) %>%
  relocate(n_events)

df_converg_mutations %>%
  filter(n_events > 1) %>%
  distinct(pair_id, cdhit_group, pan_gene_symbol, neb_locus_tag, neb_mutant_id, neb_product, MUTATION_SHORT) %>%
  arrange(cdhit_group)
```

We have 0 convergent positions 

Now we look at convergent *genes*: same cd hit group

```{r}
df_converg_genes <- df_snippy_prot_changes %>%
  group_by(cdhit_group) %>%
  mutate(n_events = n_distinct(pair_id)) %>%
  relocate(n_events)

t_converg_genes <- df_snippy_prot_changes %>%
  group_by(cdhit_group,cdhit_seq, nebraska_aa_seq, neb_gene, neb_product, neb_locus_tag, neb_mutant_id, pan_gene_symbol) %>%
  summarise(n_events = n_distinct(pair_id),
            mutations = str_c(unique(MUTATION_SHORT),
                              collapse = ", "),
            gene_prokka = str_c(unique(GENE),
                                collapse = ", "),
            product_prokka = str_c(unique(PRODUCT),
                                   collapse = ", ")) %>%
  arrange(desc(n_events)) %>%
  relocate(n_events, cdhit_group, pan_gene_symbol, neb_gene, gene_prokka, neb_locus_tag, neb_mutant_id, neb_product, product_prokka, mutations)

t_converg_genes %>%
  ungroup() %>%
  filter(n_events > 1) %>%
  select(n_events, cdhit_group, pan_gene_symbol, neb_locus_tag, neb_mutant_id, neb_product)%>%
  knitr::kable(row.names = T)
```

16 genes are convergents

# Convergence among special annotations: promoters, PSM and sRNA

```{r}
df_converg_special_annot <- df_snippy_pair_GC %>%
  drop_na(feature_type) %>%
  filter(feature_type != "NCTC8325_operons") %>%
  group_by(feature_type, FEATURE_ID) %>%
  mutate(n_events = n_distinct(pair_id)) %>%
  relocate(n_events, feature_type, FEATURE_ID) %>%
  arrange(desc(n_events)) 
```


```{r}
t_converg_special_annot <- df_converg_special_annot %>%
  group_by(n_events, feature_type, FEATURE_ID, NCTC8325_chrom, START_promoter, END_promoter, promoter_seq) %>%
  summarise(MUTATION = str_c(EVIDENCE, collapse = "/"))
```

# Convergence among special annotations: operons

```{r}
df_converg_operons <- df_snippy_pair_GC %>%
  drop_na(operon_id) %>%
  group_by(operon_id) %>%
  mutate(n_events = n_distinct(pair_id)) %>%
  relocate(n_events, operon_id, FEATURE_ID) %>%
  arrange(desc(n_events)) 
```


```{r}
t_converg_operons <- df_converg_operons %>%
  group_by(n_events, operon_id, operon_label = Names.TUshort, operon_locus_tags = Locustags.TUshort, NCTC8325_chrom, BeginTU,EndTUshort) %>%
  summarise(mutated_features = str_c(unique(FEATURE_ID), collapse = ","),
            mutated_features_types = str_c(str_sort(unique(feature_type)), collapse = ","),
            mutated_genes = str_c(na.omit(unique(pan_gene_symbol)), collapse = ",")) %>%
  arrange(-n_events)
```


```{r}
tree <- ape::read.tree("Genetic_pairs_analysis_Oct_2020/raw_data/trees/all_snippy_BPH2947/iqtree.treefile") %>%
  midpoint.root() %>%
  drop.tip("Refere-ce") 
ggtree(tree, layout = "circular")

tree_sm <- tree %>%
  drop.tip(setdiff(tree$tip.label, c(df_pair_GC$iso1, df_pair_GC$iso2)))
ggtree(tree_sm)
```

Function to plot mutated gene on tree

```{r}
plot_mutated_gene_on_tree <- function(tree,
                                      metadata,
                                      gene_id){
  matrix <- metadata %>%
    filter(cdhit_group == gene_id) %>%
    group_by(iso2) %>%
    summarise(MUTATION_SHORT = str_c(unique(MUTATION_SHORT), collapse = " ")) %>%
    column_to_rownames("iso2")
  
  title <- metadata %>%
    filter(cdhit_group == gene_id) %>%
    group_by(cdhit_group, pan_gene_symbol, neb_locus_tag, neb_mutant_id) %>%
    summarise(gene_prokka = str_c(unique(GENE),
                                collapse = ", "),
            product_prokka = str_c(unique(PRODUCT),
                                   collapse = ", ")) %>%
    unite(col = "title", 
          everything(), na.rm = T, sep = "/") %>%
    .$title
  
  subtitle <- metadata %>%
    filter(cdhit_group == gene_id) %>%
    .$n_events
  
  p <- gheatmap(ggtree(tree),
                matrix,
                colnames = F, width = .1) +
    scale_fill_brewer(na.translate = F, name = "") +
    labs(title = str_c("cd hit group ", title),
         subtitle = str_c(subtitle, " independent mutations"))
  
  return(list(p,title, subtitle))
}
```


```{r}
dir <- "Genetic_pairs_analysis_Oct_2020/figures/convergent_genes_GC_no_stats/"
dir.create(dir)

for (row in 1:nrow(t_converg_genes %>% filter(n_events > 1))){
  p <- plot_mutated_gene_on_tree(tree_sm,
                                 df_converg_genes,
                                 t_converg_genes %>% filter(n_events > 1) %>% .$cdhit_group %>% .[row])
  print(p[[1]])
  # ggsave(filename = str_c(dir, p[[3]], "#", p[[2]] %>% str_replace_all("/","#"), ".png"), plot = p[[1]], device = "png")

}
```

# Save processed data

```{r}
dir <- "Genetic_pairs_analysis_Oct_2020/processed_data/convergence/"
dir.create(dir)
subdir <- str_c(dir, "GC_no_stats/")
dir.create(subdir)

 # df_converg_genes %>%
 #   saveRDS(str_c(subdir, "df_GC_convergent_genes_no_stats.Rda"))
 # t_converg_genes %>%
 #   write_csv(str_c(subdir, "GC_convergent_genes_no_stats.csv"))
 
 df_converg_special_annot %>%
  saveRDS(str_c(subdir, "df_GC_convergent_special_annot.Rda"))
t_converg_special_annot %>%
  write_csv(str_c(subdir, "GC_convergent_special_annot.csv"))

df_converg_operons %>%
  saveRDS(str_c(subdir, "df_GC_convergent_operons.Rda"))
t_converg_operons %>%
  write_csv(str_c(subdir, "GC_convergent_operons.csv"))

# df_pair_GC %>%
#   saveRDS(str_c(subdir, "df_cleaned_pair_GC_no_stats.Rda"))
```
