---
title: "Convergence analysis among monophyletic pairs with contrasting PI clinical mortality"
author: "Stefano Giulieri"
date: "09/02/2021"
output: html_document
editor_options: 
  chunk_output_type: console
---

Here we check for convergent mutations / convergently mutated genes in carefully selected monophyletic clusters within the VANANZ dataset. Each cluster consists of a single case (lethal) strain and a bunch (>=2) of control (non-lethal) strains. In the language of pairs, cases are "iso2" and controls are "iso1". To get mutations arising in lethal isolates, we run snippy within the pair, where the de novo assembly of iso1 is the reference.

Also note that the distance to iso1-iso2 is measured as cophenetic distance.

From Romain's correlation plots we know that a distance of 4e-4 roughly corresponds to 200 denovo mutations

```{r setup, include=F}
knitr::opts_chunk$set(echo = TRUE)
knitr::opts_knit$set(root.dir = here::here())
setwd(here::here())
print(paste("My working directory is:" ,here::here()))
```

```{r message=F}
library(tidyverse)
library(ggtree)
library(phytools)
rm(list = ls())
```

# Import raw data

Processed snippy output

```{r}
df_snippy <- readRDS("Genetic_pairs_analysis_Oct_2020/processed_data/snippy/denovo/snippy_denovo_mutations.Rda")
```

Clusters dataset

```{r}
df_monophy_clusters <- read_csv("Genetic_pairs_analysis_Oct_2020/processed_data/monophyletic_clusters/df_monophyletic_mortality_clusters.csv")
```

# Filter snippy output based on clusters

Check that all clusters are present in the snippy output file

Transform the clusters file in a iso1-iso2 format

```{r}
df_mortal_paired <- df_monophy_clusters %>%
  select(iso1 = sample_id, 
         iso2 = cluster_case_sample_id,
         cluster_name) %>%
  filter(iso1 != iso2)
```


```{r}
missing_pairs <- setdiff(str_c(df_mortal_paired$iso1, df_mortal_paired$iso2, sep = "-"),
                         str_c(df_snippy$iso1, df_snippy$iso2, sep = "-"))
missing_pairs

df_mortal_paired %>%
  group_by(cluster_name) %>%
  filter(any(str_c(iso1, iso2, sep = "-") %in% missing_pairs))
```

We have 4 missing pairs and two missing clusters.

They could be missing because 1) no mutations were found or 2) they were not including in the genetic pairs, due to an artifact of the distance measured using the core genome.

To check this we need the  dataset of mutation counts

```{r}
df_denovo_mutcounts <- readRDS("Genetic_pairs_analysis_Oct_2020/processed_data/snippy/denovo/snippy_denovo_mutcounts.Rda")


df_denovo_mutcounts %>%
  filter(str_c(iso1, iso2, sep = "-") %in% missing_pairs) %>%
  filter(filter_type == "n_mask")
```

Thus, 4 pairs with contrasting mortality have 0 mutations. Since mortality is a multi-factorial phenotype, it is unlikely that the contrasting phenotype is due to some undetected structural variants (unless there is a phenotypic difference at the same time)

# Preliminary analysis of 23 clusters (21 with mutations)

```{r}
df_snippy_pair_mortal <- df_mortal_paired %>%
  left_join(df_snippy) %>%
  drop_na(CHROM)

n_distinct(df_snippy_pair_mortal$cluster_name)
```

# Convergence?

First we mask intergenic mutations and synonymous mutations.

```{r}
df_snippy_pair_mortal %>%
  count(EFFTYPE)

df_snippy_prot_changes <- df_snippy_pair_mortal %>%
  drop_na(cdhit_group) %>%
  filter(!is.na(EFFTYPE) & EFFTYPE != "synonymous_variant")
```

Check for convergent *mutations*: same gene, same mutation

```{r}
df_converg_mutations <- df_snippy_prot_changes %>%
  group_by(cdhit_group, MUTATION_SHORT ) %>%
  mutate(n_events = n_distinct(cluster_name)) %>%
  relocate(n_events)

df_converg_mutations %>%
  filter(n_events > 1)
```

Check for convergent *positions*: same gene, same position

```{r}
df_converg_positions <- df_snippy_prot_changes %>%
  group_by(cdhit_group, AA_POS ) %>%
  mutate(n_events = n_distinct(cluster_name)) %>%
  relocate(n_events)

df_converg_positions %>%
  filter(n_events > 1) %>%
  distinct(cluster_name, cdhit_group, pan_gene_symbol, neb_locus_tag, neb_mutant_id, neb_product, MUTATION_SHORT) %>%
  arrange(cdhit_group)
```

We have two positions that appear to be mutated twice independently: the gene queC at position 29 (frameshift) and an acetyltransferase at position 150.

Check for convergent *genes*: same cd hit group

```{r}
df_converg_genes <- df_snippy_prot_changes %>%
  group_by(cdhit_group) %>%
  mutate(n_events = n_distinct(cluster_name)) %>%
  relocate(n_events)

t_converg_genes <- df_snippy_prot_changes %>%
  group_by(cdhit_group, nebraska_aa_seq, neb_gene, neb_product, neb_locus_tag, neb_mutant_id, pan_gene_symbol) %>%
  summarise(n_events = n_distinct(cluster_name),
            mutations = str_c(unique(MUTATION_SHORT),
                              collapse = ", "),
            gene_prokka = str_c(unique(GENE),
                                collapse = ", "),
            product_prokka = str_c(unique(PRODUCT),
                                   collapse = ", ")) %>%
  arrange(desc(n_events)) %>%
  relocate(n_events, pan_gene_symbol, neb_gene, gene_prokka, neb_locus_tag, neb_mutant_id, neb_product, product_prokka, mutations)

t_converg_genes %>%
  ungroup() %>%
  filter(n_events > 1) %>%
  select(n_events, pan_gene_symbol, neb_locus_tag, neb_mutant_id, neb_product)%>%
  knitr::kable(row.names = T)
```

Plot these convergent mutated genes on the tree

```{r}
tree <- ape::read.tree("Genetic_pairs_analysis_Oct_2020/raw_data/trees/all_snippy_BPH2947/iqtree.treefile") %>%
  midpoint.root() %>%
  drop.tip("Refere-ce") 
ggtree(tree, layout = "circular")

tree_sm <- tree %>%
  drop.tip(setdiff(tree$tip.label, df_monophy_clusters$sample_id))
ggtree(tree_sm)
```

Function to plot mutated gene on tree

```{r}
plot_mutated_gene_on_tree <- function(tree,
                                      metadata,
                                      gene_id){
  matrix <- metadata %>%
    filter(cdhit_group == gene_id) %>%
    group_by(iso2) %>%
    summarise(MUTATION_SHORT = str_c(unique(MUTATION_SHORT), collapse = " ")) %>%
    column_to_rownames("iso2")
  
  title <- metadata %>%
    filter(cdhit_group == gene_id) %>%
    group_by(cdhit_group, pan_gene_symbol, neb_locus_tag, neb_mutant_id) %>%
    summarise(gene_prokka = str_c(unique(GENE),
                                collapse = ", "),
            product_prokka = str_c(unique(PRODUCT),
                                   collapse = ", ")) %>%
    unite(col = "title", 
          everything(), na.rm = T, sep = "/") %>%
    .$title
  
  subtitle <- metadata %>%
    filter(cdhit_group == gene_id) %>%
    .$n_events
  
  p <- gheatmap(ggtree(tree),
                matrix,
                colnames = F, width = .1) +
    scale_fill_brewer(na.translate = F, name = "") +
    labs(title = str_c("cd hit cluster ", title),
         subtitle = str_c(subtitle, " independent mutations"))
  
  return(p)
                                      }
```


```{r}
dir <- "Genetic_pairs_analysis_Oct_2020/figures/convergent_genes_mortality/"
dir.create(dir)

for (i in t_converg_genes %>% filter(n_events > 1) %>% arrange(n_events) %>% .$cdhit_group){
  p <- plot_mutated_gene_on_tree(tree_sm, 
                                 df_converg_genes,
                                 i)
  
  
  # ggsave(str_c(dir, "cd_hit_group_", i, ".pdf"))
  
}
```

# Assess convergence from snippy on BPH2947

To do

# Save processed data

```{r}
dir <- "Genetic_pairs_analysis_Oct_2020/processed_data/convergence/"
dir.create(dir)
subdir <- str_c(dir, "mortality/")
dir.create(subdir)

# df_converg_genes %>%
#   saveRDS(str_c(subdir, "df_mortality_convergent_genes.Rda"))
# t_converg_genes %>%
#   write_csv(str_c(subdir, "mortality_convergent_genes.csv"))
```