---
title: "Creation of a metadata file for the VANANZ collection"
author: "Stefano Giulieri"
date: "22/01/2020"
output: html_document
editor_options: 
  chunk_output_type: inline
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
knitr::opts_knit$set(root.dir = here::here())
msg <- stringr::str_c("My directory is ", here::here())
print(msg)
```

# Library

```{r message=FALSE}
library(tidyverse)
library(magrittr)
```

# Remove objects from the working environment

```{r}
# rm(list = ls())
```

# Outline of the metadata file

The file will provide minimal metadata for the 843 strains in the VANANZ collection. The metadata include: 
- genomic data for all strains: sequence type (ST), clonal complex (CC, to be inferred using a clustering method like CD-hit), presence of the mecA gene
- genomic data for the pairs: distance to the index, isogenic strain, sample type, mutations
- clinical data for the pairs: clinical context of the sample collection
- clinical data for the episodes: mortality at 30 days

# Sources

Raw metadata can be found in the dedicated folder.

```{r}
dir <- "plate_info/raw_metadata"
if (!dir.exists(dir)) dir.create(dir)
```

# Genomic data for all strains

This data are extracted from an existing file that it not available in the collection. However, the code is recorded here for reproducibility

```{r message=FALSE}
file <- "../VANESSA_ANZCOSS/data/VAN_VSS_n846_mlst.tab"
mlst_dat <- read_tsv(file) %>%
  mutate(sample_id = str_remove(FILE, "/contigs.fa")) %>%
  select(sample_id, ST)
file <- "../VANESSA_ANZCOSS/dataframes/resistome_n843.csv"
resistome_dat <- read_csv(file) %>%
  select(sample_id = mdu_id, mecA)
  
```

# Clinical data for the episodes

```{r message=FALSE}
file <- "../VANESSA_ANZCOSS/dataframes/pop_structure_n747_clinical_outcomes.csv"
outcomes_dat <- read_csv(file) %>%
  select(sample_id = mdu_id, mortality = outcome30d)
```

# Genomic and clinical data of the pairs

```{r message=FALSE}
file <- "../VANESSA_ANZCOSS/dataframes/whost_n130_df_new.csv"
pair_data <- read_csv(file) %>%
  select(sample_id = mdu_id, strain_group = studyid, intrahost_index, 
         included, unrelated_to_index, colonizing, intrahost_isogenic, mut_count, scv, vanco_difference, 
         persister_type, intrahost_sampledelay) %>%
  mutate(sample_type = case_when(
    intrahost_index == 1 ~ "index",
    intrahost_isogenic == 1 ~ "paired-invasive",
    colonizing == 1 ~ "paired-colonising"
  )) %>%
  # need to manually correct the scv variable
  mutate(scv = sample_id %in% c("BPH2729", "BPH2738", "BPH3733")) %>%
  select(-c(colonizing, intrahost_isogenic))

# add more mutation data
file <- "../VANESSA_ANZCOSS/dataframes/variants_n182.csv"
pair_mutations <- read_csv(file) %>%
  filter(eff_category %in% c("missense", "nonsense/frameshift")) %>%
  transmute(sample_id = ISOLATE,
           label = if_else(!is.na(GENE),
                         GENE,
                         PRODUCT),
           MUTATION = MUTATION) 
pair_genes <- pair_mutations %>%
  group_by(sample_id) %>%
  summarise(genes = str_c(unique(label), collapse = " / "))

# check numbers
nrow(pair_data %>% filter(included))
pair_data %>% filter(included) %>% .$strain_group %>% n_distinct()
```


# Create metadata file

```{r message=FALSE}
df_metadata <- read_csv("plate_info/sample_list_plate.csv") %>%
  select(sample_id = mdu_id, strain_group = studyid) %>%
  left_join(mlst_dat) %>%
  left_join(resistome_dat) %>%
  left_join(outcomes_dat) %>%
  left_join(pair_data) %>%
  left_join(pair_genes)
df_metadata %>%
  write_csv("plate_info/strain_metadata.csv")
```

