---
title: 'Summarise GWAS: PI AUC (mutations)'
author: "Stefano Giulieri"
date: "24/05/2022"
output: html_document
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
setwd(stringr::str_c(here::here(), "/GWAS/pyseer_output"))
msg <- glue::glue("My directory is {getwd()}")
message(msg)
```

Here we summarise mutation GWAS results for PI uptake (and generate plots)

```{r message=FALSE}
library(tidyverse)
library(patchwork)
library(ggrepel)
rm(list = ls())
```

# Load raw data

# Pyseer output directories

Here we focus on pyseer-gemma output. This can extended later

```{r}
f <- "~/Documents/Transfer_with_server/nomaf.nosyn.core90.mutations.toxicity_norm.pyseer.tab"
dir <- "raw_data/mutations"
file.copy(f, dir)
```

Generate a df of pyseer files

```{r}
files <- list.files(dir, recursive = T, full.names = T)
df_files <- tibble(path = files,
                   dataset = "VANANZ",
                   phenotype = str_remove(basename(files), ".pyseer.tab")) %>%
  mutate(phenotype = str_remove(phenotype, ".*mutations.|.*genes."))

df_files <- drop_na(df_files, dataset) %>%
  mutate(across(everything(), as.factor)) 
```

Parse files and concatenate them

```{r}
names(files) <- files
df_pyseer <- purrr::map_df(.x = files, .f = read_tsv, .id = "path") 

df_pyseer <- df_pyseer %>%
  mutate(dir = str_remove(dirname(path), "/raw_data/mutations"))
dirs <- unique(df_pyseer$dir)

df_pyseer <- df_pyseer %>%
  left_join(df_files)
```

## Annotation files (mutations)

```{r}
df_annotation <- readRDS("processed_data/annotation/df_annotation.Rda")
```

## Annotation files (genes)

```{r}
df_annotation_simple <- readRDS("~/OneDrive - The University of Melbourne/R/SAUREUS-GENERAL/ref_genomes/processed_data/BPH2947/df_blastp_BPH2947_neb_curated.Rda")
```


## Annotate output

```{r}
df_pyseer_annotated <- df_pyseer %>%
  dplyr::rename(mutation_id = variant,
         p_value = `lrt-pvalue`) %>%
  left_join(df_annotation) %>%
  arrange(p_value)
```

# Plot for figure 3 

Manhattan plots

```{r}
df_plot <- df_pyseer_annotated %>%
  replace_na(list(unique_gene_symbol = "intergenic")) %>%
  mutate(unique_gene_symbol = if_else(str_detect(unique_gene_symbol, "USA|EW033"),
                                      "hypothetical",
                                      unique_gene_symbol)) %>%
  mutate(mutation_label = str_c("var", POS, "_", unique_gene_symbol)) %>%
  group_by(homoplasy, rare_mutation) %>%
  mutate(bonf = .05/n_distinct(mutation_id)) %>%
  select(mutation_id, mutation_label, POS, unique_gene_symbol,  dataset, p_value, bonf, homoplasy, rare_mutation)

df_plot %>%
  ggplot(aes(x = POS, y = -log10(p_value))) +
  geom_point() +
  geom_hline(aes(yintercept = -log10(bonf)), linetype = "dotted", colour = "red") +
  facet_grid(homoplasy~rare_mutation) +
  theme_bw()

# ggsave("figures/vanco_mic/vanco_gwas_mutations_manhattan_plot.png")
```

Generate table with most signficant/interesting outputs

```{r}
# table_vanco <- df_pyseer_annotated %>%
#   select(mutation_id, POS, unique_gene_symbol, neb_product, TYPE, MUTATION_SHORT, dataset, cohort, p_value, af,  beta) %>%
#   distinct() %>%
#   group_by(dataset) %>%
#   mutate(bonf = .05/n_distinct(mutation_id),
#          rank = row_number()) %>%
#   filter(p_value <= bonf) %>%
#   relocate(rank)
# 
# prepare_table_vanco <- table_vanco %>%
#   mutate(cohort = fct_recode(cohort, `Combined` = "A + B")) %>%
#   ungroup() %>%
#   select(mutation_id, TYPE, POS, unique_gene_symbol, neb_product, MUTATION_SHORT, cohort, rank, p_value, beta) %>%
#   pivot_wider(names_from = cohort, values_from = c(rank, p_value, beta))%>%
#   mutate(cohort_A = str_c(p_value_A, " (", beta_A, ")")) %>%
#   mutate(cohort_B = str_c(p_value_B, " (", beta_B, ")")) %>%
#   mutate(combined = str_c(p_value_Combined, " (", beta_Combined, ")")) %>%
#   select(TYPE, POS, unique_gene_symbol, neb_product, MUTATION_SHORT, cohort_A, cohort_B, combined, starts_with("rank")) 
# 
# prepare_table_vanco %>%
#   rowwise() %>%
#   mutate(min_rank = min(rank_A, rank_B, rank_Combined, na.rm = T)) %>%
#   filter(min_rank <= 3 |
#            unique_gene_symbol %in% c("mprF", "pbp2") |
#            POS == 1458535) %>%
#  arrange(min_rank) %>%
#   select(-min_rank) %>%
#   write_csv("processed_data/prepare_table_vanco_mutations.csv")
``` 

# Save processed output

```{r}
# df_pyseer_annotated %>%
#   saveRDS("processed_data/vanco_phenotype/pyseer_output_vanco_mutations.Rda")
```

