---
title: "Compile table of genetic_pairs and mutation lists: mapping on the de novo assembly of iso1"
author: "Stefano Giulieri"
date: "17/02/2021"
output: html_document
---

```{r setup}
knitr::opts_chunk$set(echo = TRUE)
knitr::opts_knit$set(root.dir = here::here())
msg <- glue::glue("My directory is {here::here()}")
message(msg)
```

This script focuses on processing the output of snippy run on genetic pairs, where reads were mapped on iso1 (= internal reference)

```{r message=F}
library(tidyverse)
rm(list = ls())
```

# Raw data

List with all (current) genetic pairs run on snippy (from the server)

```{r}
f <- "~/Documents/Transfer_with_server/genetic_pairs.tab"
dir <- "Genetic_pairs_analysis_Oct_2020/raw_data/pairs_metadata/"
fname <- basename(f)
# file.copy(f,
#           str_c(dir, fname),
#           overwrite = T)
df_pairs <- read_tsv(f, 
                     col_names = c("PAIR_ID", 
                                   "iso1",
                                   "iso2"))

```

Snippy output

```{r}
dir <- "Genetic_pairs_analysis_Oct_2020/raw_data/snippy/denovo/"
dir.create(dir)
f <- "~/Documents/Transfer_with_server/all_snps.mask.tab"
fname <- basename(f)
# file.copy(f, str_c(dir, fname), overwrite = T)
snippy_denovo <- read_tsv(str_c(dir, fname)) %>%
  arrange(PAIR_ID) 
glimpse(snippy_denovo)
```

# Edit snippy output

```{r}
# modify snippy output: 
# 1) generate iso1 and iso2 for merging with the phenotypic analysis
# 2) modify CHROM (contig name) to be consistent with the labelling in bed files down the track
# 3) extract mutation effects for easier interpretation


source("Functions/all_functions.R")
snippy_denovo_modified <- snippy_denovo %>%
  mutate(iso1 = str_remove(REFERENCE, ".gbk"),
         iso2 = ISOLATE,
         CHROM = str_c(iso1, "_", CHROM)) %>%
  separate(EFFECT, 
           into = c("EFFTYPE", "NUCLEOTIDE_CHANGE", "MUTATION"), 
           sep = "\\s", 
           remove = T, 
           extra = "merge") %>%
  mutate(NUCLEOTIDE_CHANGE = str_remove(NUCLEOTIDE_CHANGE, "c."),
         MUTATION = str_remove(MUTATION, "p."),
         MUTATION_SHORT = aa_convert(MUTATION)) %>%
  separate(AA_POS, 
           into = c("AA_POS", "AA_LENGTH"), 
           sep = "/", 
           remove = T) %>%
  mutate_at(vars(starts_with("AA")), 
            as.numeric)
glimpse(snippy_denovo_modified)

snippy_denovo_modified %>%
  .$ISOLATE %>%
  n_distinct()

rm(snippy_denovo)
```

# clustering of protein genes (cd-hit)

cd-hit was first run on the amino acid sequences of the mutated genes

cd-hit was run using default parameters

id threshold (`-C`) 0.9, this is defined as number identical amino acids or bases in alignment divided by the full length of the shorter sequence
the alignment coverage controls (-aL, -AL, -aS, -AS) defaults are 0

```{r}
f <- "~/Documents/Transfer_with_server/mutated_proteins.cd-hit.tab"
fname <- basename(f)
# file.copy(f, str_c(dir, fname), overwrite = T)

protein_clusters_data <- read_tsv(str_c(dir, fname)) %>%
  group_by(clstr) %>%
  mutate(clstr_rep_id = id[which(clstr_rep == 1)]) %>%
  select(LOCUS_TAG = id, clstr_group = clstr, clstr_size, LOCUS_TAG_aa_length = length, clstr_iden, clstr_cov, clstr_rep_id) %>%
  rename_with(~ str_replace(.x, "clstr", "cdhit"))
glimpse(protein_clusters_data)

f <- "~/Documents/Transfer_with_server/mutated_proteins.cd-hit.representative.tab"
fname <- basename(f)
# file.copy(f, str_c(dir, fname), overwrite = T)

protein_seq_data <- read_tsv(str_c(dir, fname), col_names = c("cdhit_rep_id", "cdhit_seq")) %>%
  left_join(protein_clusters_data %>% distinct(cdhit_rep_id, cdhit_group))
glimpse(protein_seq_data)
```

Join to snippy output

```{r}
snippy_denovo_modified_proteins <- snippy_denovo_modified %>%
  left_join(protein_clusters_data) %>%
  left_join(protein_seq_data)

# check that each LOCUS TAG has a cd hit group
snippy_denovo_modified_proteins %>%
  drop_na(LOCUS_TAG) %>%
  filter(is.na(cdhit_group)) 

snippy_denovo_modified_proteins %>%
  drop_na(LOCUS_TAG) %>%
  filter(is.na(cdhit_group)) %>%
  distinct(GENE, PRODUCT)

# We have 330 observations with existing LOCUS TAG and missing cd hit group. They are all ribosomal genes

rm(snippy_denovo_modified,
   protein_clusters_data,
   protein_seq_data)
```

# Add FPR3757 annotation

We have used both cd-hit and blastp
We select the blastp annotation

```{r}
f <- "~/Documents/Transfer_with_server/mutated_proteins_FPR3757_blastp.tab"
fname <- basename(f)
# file.copy(f, str_c(dir, fname), overwrite = T)

blastp_data <- read_tsv(str_c(dir, fname),
                        col_names = c("QUERY", "SUBJECT", "PIDENT", "ALIGNLEN", "MISMATCH", "GAPS", "QSTART", "QEND", "SSTART", "SEND", "EVALUE", "BITSCORE", "QLEN"))
glimpse(blastp_data)

df_blastp <- blastp_data %>%
  mutate(PCOV = ALIGNLEN/QLEN*100) # we need to calculate the coverge here because the qcov* variables in blastp don't represent coverage as calculated here (checked manually looking at local alignments)

df_blastp %>%
  ggplot(aes(x = PCOV)) +
  geom_histogram() +
  theme_bw()

df_blastp %>%
  ggplot(aes(x = PIDENT)) +
  geom_histogram() +
  theme_bw()

df_blastp <- df_blastp %>%
  filter(PIDENT >=90 & PCOV >= 50) %>%
  arrange(QUERY, desc(PIDENT, PCOV)) %>%
  group_by(QUERY) %>%
  slice_head(n = 1) %>%
  select(cdhit_rep_id = QUERY, neb_locus_tag = SUBJECT, cdhit_rep_aa_length = QLEN, PIDENT, PCOV)

rm(blastp_data)

snippy_denovo_modified_proteins <- snippy_denovo_modified_proteins %>%
  left_join(df_blastp)

```

Add  NEB mutants id

```{r}
df_neb <- read_csv("Ideas_Grant_2020_analysis/Raw_data/nebraska_all_proteins.csv") %>%
  rename(neb_mutant_id = `Strain Name`) %>%
  select(-c(`Gene name`, `gene discription`))
glimpse(df_neb)

# check for duplicates
df_neb %>%
  group_by(nebraska_locus_tag) %>%
  filter(n() > 1)

# fix this
df_neb <- df_neb %>%
  group_by(nebraska_locus_tag) %>%
  mutate(neb_comment = if_else(n() > 1,
                               str_c("There are more than one NEB mutants for ", unique(nebraska_locus_tag), ": ", str_c(neb_mutant_id, collapse = ", ")),
                               NA_character_)) %>%
  filter(row_number() == 1)

snippy_denovo_modified_proteins <- snippy_denovo_modified_proteins %>%
  left_join(df_neb, by = c("neb_locus_tag" = "nebraska_locus_tag"))
```

Add aureowiki

```{r}
df_aureowiki <- read_csv("Ideas_Grant_2020_analysis/Genetic_pairs_analysis_completed/Raw_data/aureowiki_USA300_FPR3757.csv", guess_max = 3000) 
glimpse(df_aureowiki)
# Fix problematic names and reduce number of variables (for now)
df_aureowiki <- df_aureowiki %>%
  select(neb_locus_tag = `locus tag`,
         neb_locus_tag2 = `new locus tag`,
         pan_gene_symbol = `pan gene symbol`,
         pan_locus_tag = `pan locus tag`,
         neb_product = product,
         neb_gene = symbol,
         neb_strand = strand,
         neb_start = start,
         neb_end = end,
         neb_gene_length = `gene length`,
         MicrobesOnline, BioCyc,
         neb_operon = operon
         )

# check for duplicates
df_aureowiki %>%
  group_by(neb_locus_tag) %>%
  filter(n() > 1)

snippy_denovo_modified_proteins <- snippy_denovo_modified_proteins %>%
  left_join(df_aureowiki)

```

# Add promoter annotation (TSS Emote)

```{r}
f <- "~/Documents/Transfer_with_server/all_mutated_intergenic_promoters.bed"
fname <- basename(f)

# file.copy(f, str_c(dir, fname), overwrite = T)

promoter_data <- read_tsv(str_c(dir, fname), col_names = c("CHROM_prom", "START_prom", "END_prom", "promoter_id", "score", "STRAND_prom", "SSTART", "SSEND", "CHROM", "START", "END", "PAIR_ID" )) %>%
   select(PAIR_ID, chrom_join = CHROM, POS = END, promoter_id, promoter_strand = STRAND_prom, promoter_start = START_prom, promoter_end = END_prom ) 

# snippy_denovo_modified_proteins <- snippy_denovo_modified_proteins %>%
#  
#   left_join(promoter_data)
# 
# snippy_denovo_modified_proteins %>%
#   drop_na(promoter_id)
```

# Add special annotation (NCTC8325, Maeder, PLoS Genet 2016)

```{r}
f <- "~/Documents/Transfer_with_server/all_snps.features.bed"
fname <- basename(f)
file.copy(f, str_c(dir, fname), overwrite = F)

special_annot <- read_tsv(str_c(dir, fname),
                          col_names = c("CHROM_f", "START_feature", "END_feature", "FEATURE_ID", "SCORE", "STRAND", "SSTART", "SEND", "CHROM", "POS_minus_1", "POS", "PAIR_ID"))%>%
  mutate(FEATURE_ID = str_split_fixed(FEATURE_ID, "::", 2)[,1])

special_annot <- special_annot %>%
  select(PAIR_ID, CHROM, POS, FEATURE_ID, START_feature, END_feature)

f <- "~/Documents/Transfer_with_server/special_features.tab"
fname <- basename(f)
file.copy(f, str_c(dir, fname), overwrite = F)

special_annot_files <- read_tsv(str_c(dir, fname),
                          col_names = c("features_file", "FEATURE_ID")) %>%
  mutate(FEATURE_ID = str_split_fixed(FEATURE_ID, "::", 2)[,1]) %>%
  mutate(feature_type = case_when(
    str_detect(features_file, "promoters") ~ "NCTC8325_promoter",
    str_detect(features_file, "psm") ~ "PSM",
    str_detect(features_file, "sRNA") ~ "sRNA",
    str_detect(features_file, "features") ~ "NCTC8325_operons",
    str_detect(features_file, "Emote") ~ "TSS_EMOTE_promoter"
  ))

snippy_annotated <- snippy_denovo_modified_proteins %>%
  mutate(chrom_join = str_split_fixed(CHROM, "_", 2)[,2]) %>%
  left_join(special_annot, by = c("PAIR_ID", "chrom_join" = "CHROM", "POS")) %>%
  left_join(special_annot_files) %>%
  select(-chrom_join, -features_file)

# View(snippy_annotated %>% filter(PAIR_ID == "GP-3022" & POS == 27787))

```

Add more annotation from the Maeder, PLoS genet 2016 paper and from the TSS-EMOTE paper

```{r}
dir <- "Genetic_pairs_analysis_Oct_2020/raw_data/snippy/"
f <- "../SAUREUS-GENERAL/ref_genomes/raw_data/S5Table_promoters.xlsx"
fname <- basename(f)
file.copy(f, str_c(dir, fname), overwrite = F)

promoters_table <- readxl::read_xlsx(f, skip = 1) %>%
  transmute(FEATURE_ID = Id, operon_id = str_c("TU-", FEATURE_ID), promoter_seq = SigmaFactorBS, Names.TUshort, Locustags.TUshort, BeginTU, EndTUshort) 

f <- "~/Documents/Transfer_with_server/Sa_NCTC8325_promoters.bed"
fname <- basename(f)
file.copy(f, str_c(dir, fname), overwrite = F)
promoters_coord <- read_tsv(str_c(dir, fname), 
                            col_names = c("CHROM", "START_promoter", "END_promoter", "FEATURE_ID", "SCORE", "STRAND", "SSTART", "SEND"), 
                            skip = 1) %>%
  select(FEATURE_ID, START_promoter, END_promoter)

promoters_table <- promoters_table %>%
  left_join(promoters_coord)

dir <- "Genetic_pairs_analysis_Oct_2020/raw_data/snippy/"
f <- "../SAUREUS-GENERAL/ref_genomes/raw_data/S6Table_features.xlsx"
fname <- basename(f)
file.copy(f, str_c(dir, fname), overwrite = F)

operons_table <- readxl::read_xlsx(f, skip = 1) %>%
  separate_rows(TUshortUp, sep = ", ") %>%
  transmute(FEATURE_ID = Locustag, operon_feature_name = Name, operon_feature_start = Start, operon_feature_end = End, TUshortUp, operon_id = str_c("TU-", TUshortUp)) 

# add coordinates of the transcription unit
operons_table <- operons_table %>%
  left_join(promoters_table %>% select(operon_id, Names.TUshort:EndTUshort))

# TSS-Emote annotations


df_annotations <- operons_table %>%
  full_join(promoters_table) %>%
  mutate(NCTC8325_chrom = "NC_007795")

snippy_annotated <- snippy_annotated %>%
  left_join(df_annotations)
```

Add sRNA coordinates

```{r}
f <- "~/Documents/Transfer_with_server/sRNA_USA300_FPR3757.bed"
fname <- basename(f)
file.copy(f, str_c(dir, fname), overwrite = F)

sRNA_coord <- read_tsv(str_c(dir, fname), 
                            col_names = c("CHROM", "START_sRNA", "END_sRNA", "FEATURE_ID", "SCORE", "STRAND", "SSTART", "SEND"), 
                            skip = 1) %>%
  transmute(FEATURE_ID, FPR3757_chrom =  "CP000255",START_sRNA, END_sRNA) 
```


# Export processed data


```{r}
dir <- "Genetic_pairs_analysis_Oct_2020/processed_data/snippy/"
dir.create(dir)

# snippy denovo
subdir <- str_c(dir, "denovo/")
dir.create(subdir)
# snippy_denovo_modified_proteins %>%
#   saveRDS(str_c(subdir, "snippy_denovo_mutations_no_special_features.Rda"))
snippy_annotated %>%
  saveRDS(str_c(subdir, "snippy_annotated.Rda"))

# cdhit sequences
# protein_seq_data %>%
#   write_csv(str_c(subdir, "cdhit_seq.csv"))
```

