---
title: 'Summarise GWAS: PI AUC (mutations) / genetic pairs / matrix of mutated genes'
author: "Stefano Giulieri"
date: "02/03/2023"
output: html_document
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
setwd(stringr::str_c("~/Documents/Github/InToxSa", "/8-eLife_revision/Genetic_pairs"))
msg <- glue::glue("My directory is {getwd()}")
message(msg)
```

Here we summarise gene-burden GWAS results for PI uptake (and generate plots), however we limit the analysis to isolates in the 28 genetic pairs and limit the genotype to a matrix of mutated/non mutated FPR3757 genes within pairs

```{r message=FALSE}
library(tidyverse)
library(patchwork)
library(ggrepel)
rm(list = ls())
```

# Phenotype file

```{r}
df_pheno <- readRDS("processed_data/GWAS/df_pheno.Rda")
```

# Genotype matrix

```{r}
df_mutations <- readRDS("processed_data/gene_convergence/df_changes_multi_annot_for_gene_converg.Rda")
# df_mutations2 <- readRDS("processed_data/gene_convergence/df_PI_convergent_genes.Rda")
unique(df_mutations$EFFTYPE_SHORT)

df_geno <- df_mutations %>%
  select(sample_id = iso2, neb_locus_tag, EFFTYPE_SHORT) %>%
  distinct() %>%
  mutate(is_mutated = 1) %>% 
  full_join(tibble(sample_id = unique(df_mutations$iso1))) %>%
  arrange(sample_id)

geno <- df_geno %>%
  select(-EFFTYPE_SHORT) %>%
  distinct() %>%
  pivot_wider(names_from = "sample_id", values_from = "is_mutated", values_fill = 0) %>%
  drop_na(neb_locus_tag)

geno %>% 
  write_tsv("~/Documents/Transfer_with_server/genetic_pairs_mutated_genes.Rtab")
```

We can also try giving different values for substitutions vs truncations

```{r}
geno2 <- df_geno %>%
  filter(EFFTYPE_SHORT == "non-synonymous") %>%
  select(-EFFTYPE_SHORT) %>%
  distinct() %>%
  pivot_wider(names_from = "sample_id", values_from = "is_mutated", values_fill = 0) %>%
  drop_na(neb_locus_tag)

geno2 %>% 
  write_tsv("~/Documents/Transfer_with_server/genetic_pairs_mutated_genes_nonsyn.Rtab")

geno3 <- df_geno %>%
  filter(EFFTYPE_SHORT == "truncating") %>%
  select(-EFFTYPE_SHORT) %>%
  distinct() %>%
  pivot_wider(names_from = "sample_id", values_from = "is_mutated", values_fill = 0) %>%
  drop_na(neb_locus_tag)

geno3 %>% 
  write_tsv("~/Documents/Transfer_with_server/genetic_pairs_mutated_genes_trunc.Rtab")
```

# Pyseer output

```{r}
files <- list.files("~/Documents/Transfer_with_server/", pattern = "pres.toxicity", full.names = T)
dir <- "raw_data/pyseer_output/snippy_mask/"
for (f in files) file.copy(f, dir)
```

Generate a df of pyseer files

```{r}
files <- list.files(dir, recursive = T, full.names = T)
df_files <- tibble(path = files,
                   dataset = "Genetic pairs",
                   phenotype = str_remove(basename(files), ".pyseer.tab")) %>%
  mutate(phenotype = str_remove(phenotype, "[.*mutations.|.*genes.|*.pres.]"))

df_files <- drop_na(df_files, dataset) %>%
  mutate(across(where(~ !is.logical(.)), as.factor)) 
```

Parse files and concatenate them

```{r}
names(files) <- files
df_pyseer <- purrr::map_df(.x = files, .f = read_tsv, col_types = cols(.default = "c"), .id = "path") %>%
  mutate(across(af:intercept, as.numeric))

df_pyseer <- df_pyseer %>%
  left_join(df_files)
```

## Annotation files (genes)

```{r}
df_annotation_simple <- readRDS("~/OneDrive - The University of Melbourne/R/SAUREUS-GENERAL/ref_genomes/processed_data/BPH2947/df_blastp_BPH2947_neb_curated.Rda")
```

## Annotate output

```{r}
df_pyseer_annotated <- df_pyseer %>%
  dplyr::rename(neb_locus_tag = variant,
         p_value = `lrt-pvalue`) %>%
  left_join(df_annotation_simple) %>%
  arrange(p_value)
```

# Plot for figure 3

```{r}
df_plot <- df_pyseer_annotated %>%
  mutate(bonf = .05/n_distinct(LOCUS_TAG)) %>%
  mutate(label_gene_symbol = if_else(p_value < bonf, unique_gene_symbol, NA_character_)) %>%
  mutate(beta_colour = beta)%>%
  select(neb_locus_tag, unique_gene_symbol, label_gene_symbol, af, p_value, bonf, beta_colour, neb_start)

(p6 <- df_plot %>%
  ggplot(aes(x = neb_start/1e6, y = -log10(p_value))) +
  geom_point(aes(colour = beta_colour, size = af)) +
  geom_label_repel(aes(label = label_gene_symbol), fontface = "bold") +
    geom_hline(aes(yintercept = -log10(bonf)), linetype = "dotted") +
  scale_color_steps2(low = "blue", high = "red",
                     na.value = "grey90", name = "Effect size (transformed)") +
  scale_size_continuous(name = "Fraction with\nmutated gene", range = c(3,10)) +
  labs(title = "",
       x = "Gene position on FPR3757 chromosome (Mb)",
       y = "-log10(p value)") +
  theme_bw(base_size = 16) +
  theme(text = element_text(face = "bold")))
```

# Save processed output

```{r}
# df_pyseer_annotated %>%
#   saveRDS("processed_data/pyseer_output_toxicity_genes.Rda")

saveRDS(p6, "processed_data/GWAS/pyseer_output/internal_variants/pyseer_output_toxicity_genes_all_plots.Rda")
```
