---
title: "Generate dataframe of pairs information and summary plots of pairs metadata: independent pairs with no PI change"
author: "Stefano Giulieri"
date: '06/03/2023'
output: html_document
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
setwd(stringr::str_c("~/Documents/Github/InToxSa", "/8-eLife_revision/Genetic_pairs"))
msg <- glue::glue("My directory is {getwd()}")
message(msg)
```

Here we extract pairs with no change in PI uptake from iso1 to iso2 that are 1) *not already included in the 28 genetic pairs* ; 2) phylogenetically independent.

```{r}
library(tidyverse)
rm(list = ls())
```

# Raw data

Pairs data collected so far

```{r}
pairs_data <- readRDS("processed_data/PI_pairs_data/pairs_description_with_phylo_dataframe.Rda")

mutations <- readRDS("processed_data/variants_combined_annotation/snippy_denovo_annotated.Rda")
```

# Visualisations

```{r}
# Manual downsampling
df1 <- pairs_data %>%
  ungroup() %>%
  filter(!PI_pair) %>%
  slice_sample(prop = .01)
df2 <- pairs_data %>%
  ungroup() %>%
  filter(PI_pair) 
df_plot <- bind_rows(df1, df2)

# Now represent genetic distance and pheno distance on main axes
(p1 <- df_plot %>%
  ggplot(aes(x = dist_mash, y = delta_auc, shape = PI_pair, colour = log_p)) +
  geom_point() +
  stat_ellipse(aes(alpha = PI_pair), geom = "path", show.legend = T)+
    scale_alpha_discrete( name = "Gen pair", labels = c("No", "Yes")) +
  scale_shape_discrete( name = "Gen pair", labels = c("No", "Yes")) +
   scale_color_gradient(low = "blue", high = "yellow", name = "-log(p)") +
  labs(x = "Genetic distance", y  = "Difference in PI uptake") +
  theme_bw(base_size = 16) +
  theme(text = element_text(face = "bold")))
rm(df_plot, df1, df2)
```

The pairs of interest are at the centre of the left side of the plot (minimal genetic distance *and* minimal PI change)

```{r}
(p2 <- pairs_data %>%
  filter(dist_denovo <= 200 & pval >= .05 & abs(delta_auc) < 30) %>%
  ggplot(aes(x = dist_mash, y = delta_auc, shape = PI_pair, colour = log_p)) +
  geom_point() +
   scale_color_gradient(low = "blue", high = "yellow", name = "-log(p)") +
  labs(x = "Genetic distance", y  = "Difference in PI uptake") +
  theme_bw(base_size = 16) +
  theme(text = element_text(face = "bold")))
```

Now we try to define 

```{r}
df_gp_eqPI <- pairs_data %>%
  filter(dist_denovo <= 200 & pval > .05 & abs(delta_auc) < 30)
```

Exclude all pairs that share common ancestors with with 28 genetic pairs

```{r}
all_nodes <- pairs_data %>%
  filter(PI_pair) %>%
  mutate(all_nodes = list(c(mrca, path_nodes))) %>%
  .$all_nodes %>%
  unlist() %>%
  unique()
df_gp_eqPI <- df_gp_eqPI %>%
  filter(!mrca %in% all_nodes) %>%
  rowwise() %>%
  filter(!any(path_nodes %in% all_nodes))
```

Now select pairs that are independent from each other

```{r}
data_list <- list()
eqPI_nodes <- c()
# i = 1
data_list[[1]] <- df_gp_eqPI[1,]
eqPI_nodes <- unlist(c(df_gp_eqPI[1,]$mrca, df_gp_eqPI[1,]$path_nodes))
# i > 1
for (i in seq_along(df_gp_eqPI$pair_id)){
  if (i == 1) next
  else {
    df <- df_gp_eqPI[i,]
    nodes <- unlist(c(df_gp_eqPI[i,]$mrca, df_gp_eqPI[i,]$path_nodes))
    if (length(intersect(nodes, eqPI_nodes)) == 0){
      data_list[[length(data_list) + 1]] <- df
      eqPI_nodes <- c(eqPI_nodes, nodes)
    } else {
      next
    }
  }
}

df_eqPI <- bind_rows(data_list)

```

Which mutations

```{r}
df_eqPI_mutations <- df_eqPI %>%
  inner_join(mutations) %>%
  relocate(neb_gene_symbol, MUTATION_SHORT)
```

Look for convergence

```{r}
df_converg <- df_eqPI_mutations %>%
  drop_na(cdhit_group) %>%
  filter(EFFTYPE_SHORT %in% c("truncating", "non-synonymous")) %>%
  group_by(cdhit_group, neb_gene_symbol, neb_product, neb_locus_tag, neb_gene_length) %>%
  mutate(n_events = n_distinct(pair_id)) %>%
  arrange(-n_events, neb_gene_symbol) %>%
  relocate(n_events)
```

# Save processed data

```{r}
df_converg %>%
  saveRDS("processed_data/gene_convergence/PI_similar/df_PI_convergence_analysis_eqPI.Rda")
```

Dataframe with genes with >= 2 independent mutations. Exclude genes with no FPR3757 homolog and exclude ebh

```{r}
df_converg_genes <- df_converg %>%
  filter(n_events > 1) %>%
  filter(neb_gene_symbol != "ebh") %>%
  drop_na(neb_locus_tag) %>%
  distinct()

# stats
df_converg_genes %>%
  distinct(neb_gene_symbol, n_events) %>%
  ungroup() %>%
  count(n_events)

df_converg_genes %>%
   saveRDS("processed_data/gene_convergence/PI_similar/df_PI_convergent_genes_eqPI.Rda")
```
