---
title: 'Updated annotation: PROVEAN scores'
author: "Stefano Giulieri"
date: "15/06/2022"
output: 
  html_document:
    toc: TRUE
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
setwd(stringr::str_c(here::here(), "/InToxSa_paper/Genetic_pairs"))
msg <- glue::glue("My directory is {getwd()}")
message(msg)
```

Here we update the annotation to be consistent with the eLife paper. This includes adding aa new layer of annotation: PROVEAN scores for non-synonymous substitutions. 

```{r}
library(tidyverse)
rm(list = ls())
```

# Raw data: existing annotation

Existing combined annotation dataframe

```{r}
df_combined_annot <- readRDS("../../Genetic_pairs_analysis_Oct_2020/processed_data/snippy/denovo/snippy_annotated.Rda") %>%
  mutate(pair_id = paste0(iso1, "--", iso2))

efftypes <- df_combined_annot %>%
  count(EFFTYPE)
```

As compared to the eLife paper, we lack:

1. PROVEAN scores
2. Annotation of intergenic mutations
3. Structural variants

Before we proceed with PROVEAN we need to define `EFFTYPE_SHORT`

```{r}
df_combined_annot <- df_combined_annot %>%
   mutate(EFFTYPE_SHORT = case_when(
    FTYPE != "CDS"| is.na(FTYPE) | EFFTYPE == "intergenic_region" ~ "intergenic",
    EFFTYPE == "synonymous_variant" ~ "synonymous",
    str_detect(EFFTYPE, "missense_variant|inframe") ~ "non-synonymous",
    str_detect(EFFTYPE, "frameshift|stop") | EFFTYPE %in% c("start_lost", "initiator_codon_variant") ~ "truncating"
  )) %>%
  relocate(EFFTYPE_SHORT, .after = EFFTYPE)
efftypes <- df_combined_annot %>%
  count(EFFTYPE, EFFTYPE_SHORT)
```

Now we generate a table with unique non-synonymous substitutions that will help us generate the input files for PROVEAN.

We also include some variables that will help us assess the PROVEAN output (we don;t have `MUTTYPE`)

```{r}
df_provean_in <- df_combined_annot %>%
  filter(EFFTYPE_SHORT == "non-synonymous") %>%
  select(LOCUS_TAG, MUTATION_SHORT, AA_POS, AA_LENGTH) %>%
  arrange(LOCUS_TAG, AA_POS) %>%
  distinct()

n_distinct(df_provean_in$LOCUS_TAG)
```

Now we can export this list to the server

```{r}
df_provean_in %>%
  select(LOCUS_TAG, MUTATION_SHORT) %>%
  write_tsv("~/Documents/Transfer_with_server/provean_in.tab", col_names = F)
```

Save this intermediate form of the annotated dataset, while we wait for more layers of annotation and more variants

```{r}
df_combined_annot %>%
  saveRDS("processed_data/variants_combined_annotation/df_changes_multi_annotated.Rda")
```


# Process PROVEAN output

```{r}
f <- "~/Documents/Transfer_with_server/all_provean.out"
dir <- "raw_data/combined_annotation/"
file.copy(f, dir)
df_provean_out <- read_tsv(str_c(dir, basename(f)),
                           col_names = c("file", "MUTATION_SHORT", "provean_score")) %>%
  separate(col = file, into = c("LOCUS_TAG", "filename"), sep = "/") %>%
  select(-filename) %>%
  mutate(provean_prediction = if_else(provean_score > - 2.5, "neutral", "deleterious")) %>%
  mutate(provean_prediction = fct_relevel(provean_prediction, "neutral"))

df_provean_out <- df_provean_in %>%
  full_join(df_provean_out)
```

There are some missing data. Why?

```{r}
df_provean_missing <- df_provean_out %>%
  filter(is.na(provean_score))
```


Assess distribution of PROVEAN scores

```{r}
df_provean_out %>%
  ggplot(aes(x = provean_score, fill = provean_prediction)) +
  geom_histogram(binwidth = .5) +
  geom_vline(xintercept = -2.5) +
  facet_wrap(~MUTTYPE, scales = "free_y") +
  scale_fill_manual(values = c("blue", "red")) +
  theme_bw()
```

Substitutions and frameshifts/stop codons: ideally we would like to predict the functional impact of frameshifts and stop codons as well, but PROVEAN doesn't support this.

# Add PROVEAN score to the combined annotation dataframe

This annotation layer will be added to snippy output because it refers to the draft assembly rather than the FPR3757 homolog.

In addition, we create a new mutation category: non-synonymous (neutral) and non-synonymous (deleterious)

```{r}
df_combined_annot <- df_combined_annot %>%
  left_join(df_provean_out) %>%
  mutate(EFFTYPE_SHORT_2 = if_else(!is.na(provean_prediction),
                                   str_c(EFFTYPE_SHORT, " (", provean_prediction, ")"),
                                   as.character(EFFTYPE_SHORT))) %>%
  # fix missing provean scores (n=18). We will assume they are neutral
  mutate(EFFTYPE_SHORT_2 = if_else(EFFTYPE_SHORT_2 == "non-synonymous", "non-synonymous (neutral)", EFFTYPE_SHORT_2)) %>%
  mutate(EFFTYPE_SHORT_2 = as.factor(EFFTYPE_SHORT_2)) %>%
  relocate(EFFTYPE_SHORT_2, provean_score, provean_prediction, .after = EFFTYPE_SHORT)
```


# Save processed data

```{r}
dir <- "processed_data/combined_annotation/"


# df_combined_annot %>%
#   saveRDS(str_c(dir, "df_changes_multi_annotated.Rda"))
# 
# df_combined_annot %>%
#   write_csv(str_c(dir, "df_changes_multi_annotated.csv"))
```

