---
title: "Convergence analysis among monophyletic pairs with contrasting GC phenotype"
author: "Stefano Giulieri"
date: "25/02/2021"
output: html_document
editor_options: 
  chunk_output_type: console
---

Here we check for convergent mutations / convergently mutated genes in carefully created monophyletic pairs within the VANANZ dataset. Each cluster consists of a pair with high growth rate (iso1) and one with low growth (iso2).

To get mutations arising in low GC isolates (SCV-like), we ran snippy within the pair, where the de novo assembly of iso1 is the reference.

Also note that the distance to iso1-iso2 is measured as cophenetic distance.

From Romain's correlation plots we know that a distance of 4e-4 roughly corresponds to 200 denovo mutations

```{r setup, include=F}
knitr::opts_chunk$set(echo = TRUE)
knitr::opts_knit$set(root.dir = here::here())
setwd(here::here())
print(paste("My working directory is:" ,here::here()))
```

```{r message=F}
library(tidyverse)
library(ggtree)
library(phytools)
rm(list = ls())
```

# Import raw data

Processed snippy output

```{r}
df_snippy <- readRDS("Genetic_pairs_analysis_Oct_2020/processed_data/snippy/denovo/snippy_denovo_mutations.Rda")

# df_snippy <- df_snippy %>%
#   filter(str_detect(PAIR_ID, "GP-30", negate = T))
```

Dataset of contrasting PI uptake

```{r}
df_pair_GC <- read_csv("Genetic_pairs_analysis_Oct_2020/processed_data/monophyletic_clusters/genetic_pairs_significant_GC_change.csv") %>%
  mutate(non_directional_pair_id = str_c(iso1, iso2, sep = "-"))
```

# Filter snippy output based on clusters

Check that all pairs are present in the snippy output file

```{r}
missing_pairs <- setdiff(str_c(df_pair_GC$iso1, df_pair_GC$iso2, sep = "-"),
                         str_c(df_snippy$iso1, df_snippy$iso2, sep = "-"))
missing_pairs
```

These pairs could be missing because 1) no mutations were found or 2) they were not including in the genetic pairs, due to an artifact of the distance measured using the core genome.

To check this we need the  dataset of mutation counts

```{r}
df_denovo_mutcounts <- readRDS("Genetic_pairs_analysis_Oct_2020/processed_data/snippy/denovo/snippy_denovo_mutcounts.Rda") %>%
  filter(filter_type == "n_mask") %>%
  filter(!str_detect(PAIR_ID, "GP-30"))

df_denovo_mutcounts %>%
  filter(str_c(iso1, iso2, sep = "-") %in% missing_pairs)
```

Thus,  4 pairs are missing from snippy have 0 mutations. These could be interesting pairs for long reads sequencing or at least they should be checked for structural variants based on short reads data!

However, 4 other pairs need to be added to the snippy dataset

# Convergence among 22 pairs (18 with mutations)

```{r}
df_snippy_pair_GC <- df_pair_GC %>%
  left_join(df_snippy) %>%
  drop_na(CHROM)

n_distinct(df_snippy_pair_GC$non_directional_pair_id)
```

# Convergence?

First we mask intergenic mutations and synonymous mutations.

```{r}
df_snippy_pair_GC %>%
  count(EFFTYPE)
df_snippy_prot_changes <- df_snippy_pair_GC %>%
  drop_na(cdhit_group) %>%
  filter(!is.na(EFFTYPE) & EFFTYPE != "synonymous_variant")
```

Check for convergent *mutations*: same gene, same mutation

```{r}
df_converg_mutations <- df_snippy_prot_changes %>%
  group_by(cdhit_group, MUTATION_SHORT ) %>%
  mutate(n_events = n_distinct(non_directional_pair_id)) %>%
  relocate(n_events)

df_converg_mutations %>%
  filter(n_events > 1)
```


Check for convergent *positions*: same gene, same position

```{r}
df_converg_positions <- df_snippy_prot_changes %>%
  group_by(cdhit_group, AA_POS ) %>%
  mutate(n_events = n_distinct(non_directional_pair_id)) %>%
  relocate(n_events)

df_converg_positions %>%
  filter(n_events > 1) %>%
  distinct(non_directional_pair_id, cdhit_group, pan_gene_symbol, neb_locus_tag, neb_mutant_id, neb_product, MUTATION_SHORT) %>%
  arrange(cdhit_group, AA_POS)
```

Rather than true convergence, this appears to be a symmetric "allele swap" between iso1 and iso2 in two pairs (recombination?). Since it is symmetric it is unlikely to be associated with the phenotye.

```{r}
df_converg_genes <- df_snippy_prot_changes %>%
  group_by(cdhit_group) %>%
  mutate(n_events = n_distinct(non_directional_pair_id)) %>%
  relocate(n_events)

t_converg_genes <- df_snippy_prot_changes %>%
  group_by(cdhit_group, nebraska_aa_seq, neb_gene, neb_product, neb_locus_tag, neb_mutant_id, pan_gene_symbol) %>%
  summarise(n_events = n_distinct(non_directional_pair_id),
            mutations = str_c(unique(MUTATION_SHORT),
                              collapse = ", "),
            gene_prokka = str_c(unique(GENE),
                                collapse = ", "),
            product_prokka = str_c(unique(PRODUCT),
                                   collapse = ", ")) %>%
  arrange(desc(n_events)) %>%
  relocate(n_events, cdhit_group, pan_gene_symbol, neb_gene, gene_prokka, neb_locus_tag, neb_mutant_id, neb_product, product_prokka, mutations)

t_converg_genes %>%
  ungroup() %>%
  filter(n_events > 1) %>%
  select(n_events, cdhit_group, pan_gene_symbol, neb_locus_tag, neb_mutant_id, neb_product)%>%
  knitr::kable(row.names = T)
```

*Note*:

We have some interesting genes here, such as rpoB (with mutations outside of the rifampicin resistance determining region), rpoC, ausA, gltB, rsp.

Note that ebh appears twice: this is because cd hit split it into 2 separate groups. 

Plot these convergent mutated genes on the tree

```{r}
tree <- ape::read.tree("Genetic_pairs_analysis_Oct_2020/raw_data/trees/all_snippy_BPH2947/iqtree.treefile") %>%
  midpoint.root() %>%
  drop.tip("Refere-ce") 
ggtree(tree, layout = "circular")

tree_sm <- tree %>%
  drop.tip(setdiff(tree$tip.label, c(df_pair_GC$iso1, df_pair_GC$iso2)))
ggtree(tree_sm)
```

Function to plot mutated gene on tree

```{r}
plot_mutated_gene_on_tree <- function(tree,
                                      metadata,
                                      gene_id){
  matrix <- metadata %>%
    filter(cdhit_group == gene_id) %>%
    group_by(iso2) %>%
    summarise(MUTATION_SHORT = str_c(unique(MUTATION_SHORT), collapse = " ")) %>%
    column_to_rownames("iso2")
  
  title <- metadata %>%
    filter(cdhit_group == gene_id) %>%
    group_by(cdhit_group, pan_gene_symbol, neb_locus_tag, neb_mutant_id) %>%
    summarise(gene_prokka = str_c(unique(GENE),
                                collapse = ", "),
            product_prokka = str_c(unique(PRODUCT),
                                   collapse = ", ")) %>%
    unite(col = "title", 
          everything(), na.rm = T, sep = "/") %>%
    .$title
  
  subtitle <- metadata %>%
    filter(cdhit_group == gene_id) %>%
    .$n_events
  
  p <- gheatmap(ggtree(tree),
                matrix,
                colnames = F, width = .1) +
    scale_fill_brewer(na.translate = F, name = "") +
    labs(title = str_c("cd hit group ", title),
         subtitle = str_c(subtitle, " independent mutations"))
  
  return(p)
                                      }
```


```{r}
dir <- "Genetic_pairs_analysis_Oct_2020/figures/convergent_genes_GC/"
dir.create(dir)

for (i in t_converg_genes %>% filter(n_events > 1) %>% arrange(n_events) %>% .$cdhit_group){
  p <- plot_mutated_gene_on_tree(tree_sm, 
                                 df_converg_genes,
                                 i)
  
  
  ggsave(str_c(dir, "cd_hit_group_", i, ".pdf"))
  
}
```

# Save processed data

```{r}
dir <- "Genetic_pairs_analysis_Oct_2020/processed_data/convergence/"
dir.create(dir)
subdir <- str_c(dir, "GC/")
dir.create(subdir)

# df_converg_genes %>%
#   saveRDS(str_c(subdir, "df_GC_convergent_genes.Rda"))
# t_converg_genes %>%
#   write_csv(str_c(subdir, "GC_convergent_genes.csv"))
```
