---
title: "Prepare GWAS files"
author: "Stefano Giulieri"
date: "23/05/2022"
output: html_document
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
setwd(paste0(here::here(), "/InToxSa_paper/GWAS/Phenotypes"))
getwd()
```

Here we prepare the input files for the GWAS analysis of PI uptake: this refers to the phenotype file, as mutations have already been called.

We might consider other units of variation (genes presence/absence, k-mers)

```{r}
library(tidyverse)
library(bestNormalize)
library(patchwork)
rm(list = ls())
```

# Import summary PI data 

```{r}
PI_parameters <- readRDS("../../../Genetic_pairs_analysis_Oct_2020/processed_data/PI/dataframes/parameters/PI_sample_parameters.Rda")
```

We keep clinical isolates only (BPHXXXX). Also we reformat as tibble

```{r}
PI_parameters <- PI_parameters %>%
  as_tibble() %>%
  filter(str_detect(sample_id, "BPH"))

unique(PI_parameters$sample_id)
```

# Generate phenotype file

```{r}
pheno <- PI_parameters %>%
  select(sample_id, AUC_death_mean)
```

It is useful to inspect the phenotype to check for normality

```{r}
pheno %>%
  ggplot(aes(x = AUC_death_mean)) +
  geom_histogram(binwidth = 10) +
  theme_bw()

pheno %>%
  ggplot(aes(x = as.factor(1), y = AUC_death_mean)) +
  geom_violin() +
  geom_boxplot()

pheno %>%
   ggplot(aes(x = AUC_death_mean)) +
  geom_density() +
  theme_bw()
```

The distribution is not normal. It is worth trying some transformations

```{r}
x <- pheno$AUC_death_mean
norm <- bestNormalize(x)
norm

par(mar=c(1,1,1,1)) # this is necessary to fix the error: "Error in plot.new() : figure margins too large". See https://stackoverflow.com/questions/23050928/error-in-plot-new-figure-margins-too-large-scatter-plot
plot(norm)

# for each transformation visualise histogram
df_bestnorm <- tibble(
  untransformed = x,
  yeo_johnson = norm$x.t,
)
list_t <- list()
for (i in seq_along(norm$other_transforms)) {
  t <- norm$other_transforms[[i]]
  x.t <- t$x.t
  list_t[[i]] <- x.t
}
t <- bind_cols(list_t)
colnames(t) <- names(norm$other_transforms)
df_bestnorm <- df_bestnorm %>%
  bind_cols(t)
#print(names(norm$other_transforms[i]))

p <- df_bestnorm %>%
  pivot_longer(cols = everything(),
               names_to = "transform",
               values_to = "x") %>%
  ggplot(aes(x = x,
             fill = transform == "untransformed")) +
  geom_histogram(bins = 12, position = "dodge") +
   facet_wrap(~fct_relevel(transform, "untransformed"), scales = "free") +
  # facet_grid(cohort~transform, scales = "free") +
  scale_fill_manual(values = c("red", "blue"), guide = "none") +
  labs(x = "Transformed phenotype") +
  theme_bw(base_size = 16) +
  theme(text = element_text(face = "bold"))

p
 
p / wrap_elements(panel = ~plot(norm)) +
  plot_annotation(tag_levels = "A")

p2 <- as_tibble(norm$oos_preds) %>%
  pivot_longer(cols = everything(),
               names_to = "Transformation", values_to = "Normality estimate") %>%
  ggplot(aes(x = Transformation, y = `Normality estimate`)) +
  geom_boxplot(fill = "red", alpha = .5) +
  theme_bw(base_size = 16) +
  theme(text = element_text(face = "bold"))

p / p2 + 
  plot_layout(heights = c(2,1)) +
  plot_annotation(tag_levels = "A")

df_pheno_norm <- df_bestnorm %>%
  add_column(sample_id = pheno$sample_id, .before = 1)
```

# Final phenotype table to export

```{r}
pheno <- df_pheno_norm %>%
  select(sample_id, pi_auc = untransformed) %>%
  write_tsv("~/Documents/Transfer_with_server/toxicity_phenotype.tab")

pheno_norm <- df_pheno_norm %>%
  select(sample_id, pi_auc = yeo_johnson) %>%
   write_tsv("~/Documents/Transfer_with_server/toxicity_norm_phenotype.tab")
```

# Save processed phenotype data

```{r}
df_pheno_norm %>%
  saveRDS("processed_data/toxicity_phenotype_with_norm.Rda")
```

